packages/flutter/lib/src/widgets/undo_history.dart

```diff
@@ -0,0 +1,446 @@
+    widget.controller.onRedo.removeListener(redo);
+    _throttleTimer?.cancel();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Actions(
+      actions: <Type, Action<Intent>>{
+        UndoTextIntent: Action<UndoTextIntent>.overridable(context: context, defaultAction: CallbackAction<UndoTextIntent>(onInvoke: _undoFromIntent)),
+        RedoTextIntent: Action<RedoTextIntent>.overridable(context: context, defaultAction: CallbackAction<RedoTextIntent>(onInvoke: _redoFromIntent)),
+      },
+      child: widget.child,
+    );
+  }
+}
+
+/// Represents whether the current undo stack can undo or redo.
+@immutable
+class UndoHistoryValue {
+  /// Creates a value for whether the current undo stack can undo or redo.
+  ///
+  /// The [canUndo] and [canRedo] arguments must have a value, but default to
+  /// false.
+  const UndoHistoryValue({this.canUndo = false, this.canRedo = false});
+
+  /// A value corresponding to an undo stack that can neither undo nor redo.
+  static const UndoHistoryValue empty = UndoHistoryValue();
+
+  /// Whether the current undo stack can perform an undo operation.
+  final bool canUndo;
+
+  /// Whether the current undo stack can perform a redo operation.
+  final bool canRedo;
+
+  @override
+  String toString() => '${objectRuntimeType(this, 'UndoHistoryValue')}(canUndo: $canUndo, canRedo: $canRedo)';
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other))
+      return true;
+    return other is UndoHistoryValue && other.canUndo == canUndo && other.canRedo == canRedo;
+  }
+
+  @override
+  int get hashCode => Object.hash(
+        canUndo.hashCode,
+        canRedo.hashCode,
+      );
+}
+
+/// A controller for the undo history, for example for an editable text field.
+///
+/// Whenever a change happens to the underlying value that the [UndoHistory]
+/// widget tracks, that widget updates the [value] and the controller notifies
+/// it's listeners. Listeners can then read the canUndo and canRedo
+/// properties of the value to discover whether [undo] or [redo] are possible.
+///
+/// The controller also has [undo] and [redo] methods to modify the undo
+/// history.
+///
+/// {@tool dartpad}
+/// This example creates a [TextField] with an [UndoHistoryController]
+/// which provides undo and redo buttons.
+///
+/// ** See code in examples/api/lib/widgets/undo_history/undo_history_controller.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+/// * [EditableText], which uses the [UndoHistory] widget and allows
+///   control of the underlying history using an [UndoHistoryController].
+class UndoHistoryController extends ValueNotifier<UndoHistoryValue> {
+  /// Creates a controller for an [UndoHistory] widget.
+  UndoHistoryController({UndoHistoryValue? value}) : super(value ?? UndoHistoryValue.empty);
+
+  /// Notifies listeners that [undo] has been called.
+  final ChangeNotifier onUndo = ChangeNotifier();
```

Comment:
Should these ChangeNotifiers be disposed?

Reply:
Can it be private? It seems to be a way an API just to wire up the undo/redo to the UndoHistory widget.

Reply:
Possibly, but I think making it public makes this more useful to consumers, for example, if they are implementing their own version of UndoHistory, they would need to listen to the undo/redo notifications.